Assignment follow the that keypoints
-SVM,Decision Tree,Random forest and Logistic Regression
- Train-test split model training and evaluation
- Cross Validation
- Select best performing model using CV
- fine tune that model
- Evaluate again
- Apply ada boost on fine tune model

# Import Libraries

In [70]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegressionCV
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
import warnings
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import KFold
from sklearn.ensemble import AdaBoostClassifier
warnings.filterwarnings('ignore')

## Load dataset

In [2]:
df=pd.read_csv('DataSet/titanic.csv')

In [3]:
df.head() # show top 5 rows of dataset

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
df.shape # checck the shape of data

(891, 12)

In [5]:
df.info() # Check the data type of individual columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
df.describe().T # overview of data

count        mean         std   min       25%       50%    75%  \
PassengerId  891.0  446.000000  257.353842  1.00  223.5000  446.0000  668.5   
Survived     891.0    0.383838    0.486592  0.00    0.0000    0.0000    1.0   
Pclass       891.0    2.308642    0.836071  1.00    2.0000    3.0000    3.0   
Age          714.0   29.699118   14.526497  0.42   20.1250   28.0000   38.0   
SibSp        891.0    0.523008    1.102743  0.00    0.0000    0.0000    1.0   
Parch        891.0    0.381594    0.806057  0.00    0.0000    0.0000    0.0   
Fare         891.0   32.204208   49.693429  0.00    7.9104   14.4542   31.0   

                  max  
PassengerId  891.0000  
Survived       1.0000  
Pclass         3.0000  
Age           80.0000  
SibSp          8.0000  
Parch          6.0000  
Fare         512.3292

In [7]:
df.isnull().sum() # Check the missing value in dataset

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [8]:
df.isnull().sum().sort_values(ascending=False)/len(df)*100

Cabin          77.104377
Age            19.865320
Embarked        0.224467
PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
dtype: float64

**Observation**
- Cabin contains almost 77% missing value
- Age contains almost 19% missing value


In [9]:
df['Initial']=0
for i in df:
    df['Initial']=df.Name.str.extract('([A-Za-z]+)\.') #lets extract the Salutations

In [10]:
df['Initial'].replace(
    ['Mlle','Mme','Ms','Dr','Major','Lady','Countess','Jonkheer','Col','Rev','Capt','Sir','Don'],
    ['Miss','Miss','Miss','Mr','Mr','Mrs','Mrs','Other','Other','Other','Mr','Mr','Mr'],inplace=True)

In [11]:
#Assigning the missing value
df.loc[(df.Age.isnull())&(df.Initial=='Mr'),'Age']=33
df.loc[(df.Age.isnull())&(df.Initial=='Mrs'),'Age']=36
df.loc[(df.Age.isnull())&(df.Initial=='Master'),'Age']=5
df.loc[(df.Age.isnull())&(df.Initial=='Miss'),'Age']=22
df.loc[(df.Age.isnull())&(df.Initial=='Other'),'Age']=46

In [12]:
df.Age.isnull().sum()# check the missing value

0

In [13]:
df['Cabin'] = df['Cabin'].apply(lambda x:not isinstance(x, float))

In [14]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Initial'],
      dtype='object')

### Drop Columns

In [15]:
df.drop(['PassengerId','Name','Ticket'],axis=1,inplace=True) #drop the unvaluable columns

In [16]:
df.head(2) # check how many columns are left

Survived  Pclass     Sex   Age  SibSp  Parch     Fare  Cabin Embarked  \
0         0       3    male  22.0      1      0   7.2500  False        S   
1         1       1  female  38.0      1      0  71.2833   True        C   

  Initial  
0      Mr  
1     Mrs

### Label Encoding

In [17]:
label_encoder={}
for x in ['Sex','Embarked','Initial']:
    label_encoder[x]=LabelEncoder()
    df[x]=label_encoder[x].fit_transform(df[x])

In [18]:
df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Cabin  Embarked  \
0         0       3    1  22.0      1      0   7.2500  False         2   
1         1       1    0  38.0      1      0  71.2833   True         0   
2         1       3    0  26.0      0      0   7.9250  False         2   
3         1       1    0  35.0      1      0  53.1000   True         2   
4         0       3    1  35.0      0      0   8.0500  False         2   

   Initial  
0        2  
1        3  
2        1  
3        3  
4        2

**Observation**
- label encoding

## split the feature and target output

In [19]:
x=df[df.columns[1:]].values
y=df.Survived

## Using train-test split

In [20]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42,shuffle=True)

## Logistic Regression Model

In [21]:
lr=LogisticRegressionCV()
lr.fit(X_train,y_train)

LogisticRegressionCV()

In [22]:
y_pred=lr.predict(X_test)

In [23]:
print('Accuracy:',accuracy_score(y_test,y_pred))
print("Precision:",precision_score(y_test,y_pred))
print("Recall_Score:",recall_score(y_test,y_pred))
print("f1_score:",f1_score(y_test,y_pred))

Accuracy: 0.8161434977578476
Precision: 0.7857142857142857
Recall_Score: 0.7415730337078652
f1_score: 0.7630057803468209


## Decision Tree Model

In [24]:
dt=DecisionTreeClassifier()

In [25]:
dt.fit(X_train,y_train)

DecisionTreeClassifier()

In [26]:
y_pred=dt.predict(X_test)

In [27]:
print('Accuracy:',accuracy_score(y_test,y_pred))
print("Precision:",precision_score(y_test,y_pred))
print("Recall_Score:",recall_score(y_test,y_pred))
print("f1_score:",f1_score(y_test,y_pred))

Accuracy: 0.7443946188340808
Precision: 0.6818181818181818
Recall_Score: 0.6741573033707865
f1_score: 0.6779661016949152


## SVM Model

In [28]:
svm=SVC()
svm.fit(X_train,y_train)

SVC()

In [29]:
y_pred=svm.predict(X_test)

In [30]:
print('Accuracy:',accuracy_score(y_test,y_pred))
print("Precision:",precision_score(y_test,y_pred))
print("Recall_Score:",recall_score(y_test,y_pred))
print("f1_score:",f1_score(y_test,y_pred))

Accuracy: 0.6636771300448431
Precision: 0.71875
Recall_Score: 0.25842696629213485
f1_score: 0.3801652892561984


## Random Forest Model

In [31]:
rf=RandomForestClassifier()

In [32]:
rf.fit(X_train,y_train)

RandomForestClassifier()

In [33]:
y_pred=rf.predict(X_test)

In [34]:
print('Accuracy:',accuracy_score(y_test,y_pred))
print("Precision:",precision_score(y_test,y_pred))
print("Recall_Score:",recall_score(y_test,y_pred))
print("f1_score:",f1_score(y_test,y_pred))

Accuracy: 0.8161434977578476
Precision: 0.7666666666666667
Recall_Score: 0.7752808988764045
f1_score: 0.7709497206703911


# Conclusion

- **Logistic Regression** Model Show highest Accuracy

# Cross Validation using K-Folds

In [35]:
def Cross_Validation_Model(x,y,k_fold_validation=KFold(8)):
    model=LogisticRegressionCV(random_state=42)
    result1=cross_val_score(model,x,y,cv=k_fold_validation)
    print('LogisticRegression Model Kfold 8 Value')
    print(' ')
    print(result1)
    print("The Accuracy of LogisticRegressionCV :{}".format(np.mean(result1)*100))
    print('_______________________')
    
    model=DecisionTreeClassifier(random_state=42)
    result1=cross_val_score(model,x,y,cv=k_fold_validation)
    print('DecisionTreeClassifier Model Kfold 8 Value')
    print(' ')
    
    print(result1)
    print("The Accuracy of DecisionTreeClassifier :{}".format(np.mean(result1)*100))
    print('_______________________')
    
    
    model=SVC(random_state=42)
    result1=cross_val_score(model,x,y,cv=k_fold_validation)
    print('SVC Model Kfold 8 Value')
    print(' ')
    
    print(result1)
    print("The Accuracy of SVC :{}".format(np.mean(result1)*100))
    print('_______________________')
    
    model=RandomForestClassifier(random_state=42)
    result1=cross_val_score(model,x,y,cv=k_fold_validation)
    print('RandomForestClassifier Model Kfold 8 Value')
    print(' ')
    
    print(result1)
    print("The Accuracy of RandomForestClassifier :{}".format(np.mean(result1)*100))
    print('_______________________')
    

In [36]:
Cross_Validation_Model(x,y)

LogisticRegression Model Kfold 8 Value
 
[0.80357143 0.8125     0.78571429 0.82882883 0.76576577 0.76576577
 0.79279279 0.82882883]
The Accuracy of LogisticRegressionCV :79.7970962033462
_______________________
DecisionTreeClassifier Model Kfold 8 Value
 
[0.74107143 0.74107143 0.74107143 0.79279279 0.81081081 0.79279279
 0.76576577 0.84684685]
The Accuracy of DecisionTreeClassifier :77.90279118404119
_______________________
SVC Model Kfold 8 Value
 
[0.60714286 0.64285714 0.69642857 0.64864865 0.67567568 0.7027027
 0.71171171 0.67567568]
The Accuracy of SVC :67.01053732303733
_______________________
RandomForestClassifier Model Kfold 8 Value
 
[0.8125     0.83035714 0.78571429 0.83783784 0.84684685 0.81081081
 0.82882883 0.84684685]
The Accuracy of RandomForestClassifier :82.4967824967825
_______________________


**Observation**
-  In Cross-Validation Technique the **RandomForestClassifier** show optimal result

### So, I select the RandomForestClassifier and Fine Tune that Model

In [38]:
kvalue=KFold(8)
model=RandomForestClassifier(random_state=42)
result1=cross_val_score(model,x,y,cv=kvalue)

In [39]:
np.mean(result1)

0.824967824967825

# GridSearchCV

In [52]:
# numbers of tree in random forests
n_estimators=[20,60,100,120]
# number of features to considers to every split
max_features=[0.2,0.6,1.0]
# maximum number of levels in tree
max_depth=[2,8,None]
#Number of samples
max_samples=[0.5,0.75,1.0]
bootstrap = [True, False]

In [53]:
param_grid={
    "n_estimators":n_estimators,
    "max_features":max_features,
    "max_depth":max_depth,
    "max_samples":max_samples,
    "bootstrap":bootstrap
}

In [54]:
rf_grid=GridSearchCV(estimator=rf,
                    param_grid=param_grid,
                    cv=5,verbose=2,n_jobs=-1)

<IPython.core.display.Javascript object>

In [55]:
rf_grid.fit(X_train,y_train)

Fitting 5 folds for each of 216 candidates, totalling 1080 fits


GridSearchCV(cv=5, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'bootstrap': [True, False], 'max_depth': [2, 8, None],
                         'max_features': [0.2, 0.6, 1.0],
                         'max_samples': [0.5, 0.75, 1.0],
                         'n_estimators': [20, 60, 100, 120]},
             verbose=2)

In [57]:
rf_grid.best_params_

{'bootstrap': True,
 'max_depth': None,
 'max_features': 0.2,
 'max_samples': 0.5,
 'n_estimators': 100}

In [59]:
rf_grid.best_score_*100

83.53944562899788

In [62]:
rf_grid.best_score_*100-np.mean(result1)*100

1.042663132215381

**Observation**
- On Apply hyper-tuning Model performs 1.04% much better

In [63]:
def evaluate(model, X_test, y_test):
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    print('Model Performance')
    print('Accuracy: {:0.3f}'.format(accuracy*100))
    print('F1 Score: {:0.3f}'.format(f1*100))
    
    return [accuracy, f1]

In [66]:
best_grid = rf_grid.best_estimator_

grid_accuracy = evaluate(best_grid, X_test, y_test)


Model Performance
Accuracy: 82.511
F1 Score: 77.966


# AdaBoost Classifier in Python

In [87]:
abc=AdaBoostClassifier(n_estimators=50, base_estimator=rf,learning_rate=1)

In [88]:

model = abc.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = model.predict(X_test)

In [89]:
print("Accuracy:",accuracy_score(y_test,y_pred))

Accuracy: 0.7937219730941704
